blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/datastax/DAbstractMetricsRW.java

```diff
@@ -0,0 +1,259 @@
+
+                // lookup the right writer
+                RollupType rollupType = writeContext.getRollup().getRollupType();
+                DAbstractMetricIO io = getIO( rollupType.name().toLowerCase(), granularity );
+
+                ResultSetFuture future = io.putAsync(locator, writeContext.getTimestamp(), rollup, writeContext.getGranularity(), ttl);
+                futureLocatorMap.put(future, writeContext);
+            }
+
+            for (ResultSetFuture future : futureLocatorMap.keySet()) {
+                try {
+                    future.getUninterruptibly().all();
+                } catch (Exception ex) {
+                    Instrumentation.markWriteError();
+                    SingleRollupWriteContext writeContext = futureLocatorMap.get(future);
+                    LOG.error(String.format("error writing to locator %s, granularity %s", writeContext.getLocator(), writeContext.getGranularity()), ex);
+                }
+            }
+        } finally {
+            ctx.stop();
+        }
+    }
+
+    /**
+     * Fetches {@link com.rackspacecloud.blueflood.outputs.formats.MetricData} objects for the
+     * specified {@link com.rackspacecloud.blueflood.types.Locator} and
+     * {@link com.rackspacecloud.blueflood.types.Range} from the specified column family
+     *
+     * @param locator
+     * @param range
+     * @param granularity
+     * @return
+     */
+    @Override
+    public MetricData getDatapointsForRange(final Locator locator,
+                                            Range range,
+                                            Granularity granularity) throws IOException {
+        Map<Locator, MetricData> result = getDatapointsForRange(new ArrayList<Locator>() {{ add(locator); }}, range, granularity);
+        return result.get(locator);
+    }
+
+    /**
+     * Fetches {@link com.rackspacecloud.blueflood.outputs.formats.MetricData} objects for the
+     * specified {@link com.rackspacecloud.blueflood.types.Locator} and
+     * {@link com.rackspacecloud.blueflood.types.Range} from the specified column family
+     *
+     * This is a helper method used to get basic metrics as well as preaggregated metrics.  Hence, both granularity
+     * and columnFamily are required.
+     *
+     * @param locators
+     * @param range
+     * @param columnFamily
+     * @param granularity
+     * @return
+     */
+    protected Map<Locator, MetricData> getDatapointsForRange( List<Locator> locators,
+                                                              Range range,
+                                                              String columnFamily,
+                                                              Granularity granularity ) {
+        // in this loop, we will fire all the executeAsync() of
+        // various select statements, the collect all of the
+        // ResultSetFutures
+        Map<Locator, List<ResultSetFuture>> locatorToFuturesMap = new HashMap<Locator, List<ResultSetFuture>>();
```

Comment:
Never mind... it looks like the caller puts the Instrumentation... I wonder if we should change that...